File: picker/src/main/java/com/hy/picker/PictureEditPreviewActivity.java
package com.hy.picker;

import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Build.VERSION;
import android.os.Bundle;
import android.os.Looper;
import android.os.MessageQueue;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.facebook.drawee.drawable.ScalingUtils;
import com.facebook.drawee.generic.GenericDraweeHierarchy;
import com.hy.picker.utils.AttrsUtils;
import com.hy.picker.utils.CommonUtils;
import com.hy.picker.utils.DisplayOptimizeListener;
import com.hy.picker.view.ImageSource;
import com.hy.picker.view.PickerScaleImageView;
import com.picker2.model.Photo;
import com.picker2.utils.AndroidLifecycleUtils;

import java.io.File;
import java.lang.ref.WeakReference;

import androidx.core.content.ContextCompat;
import me.relex.photodraweeview.PhotoDraweeView;

/**
 * Created time : 2018/8/2 8:23.
 *
 * @author HY
 */
public class PictureEditPreviewActivity extends BaseActivity {

    private RelativeLayout mWholeView;
    private View mToolbarTop;
    private ImageView mBtnBack;
    private TextView mBtnSend;
    //    private AppCompatRadioButton mUseOrigin;
//    private PhotoView mPhotoView;
//    private BigImageView mLongIv;
    private boolean mFullScreen;

    private Photo mPicItem;
    private Drawable mDefaultDrawable;
    private View mView;

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.picker_activity_edit_preview);

        mDefaultDrawable = AttrsUtils.getTypeValueDrawable(this, R.attr.picker_image_default);
        if (null == mDefaultDrawable) {
            mDefaultDrawable = ContextCompat.getDrawable(this, R.drawable.picker_grid_image_default);
        }

        initView();

        if (VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            mToolbarTop.setPadding(0, CommonUtils.getStatusBarHeight(this), 0, 0);
        }

        Intent intent = getIntent();
        mPicItem = intent.getParcelableExtra(EXTRA_ITEM);

        if (mPicItem == null) {
            Toast.makeText(this, R.string.picker_file_error, Toast.LENGTH_SHORT).show();
            finish();
            return;
        }


        if (mPicItem.isLong()) {
            mView = new PickerScaleImageView(this);
            mView.setOnClickListener(v -> {
                mFullScreen = !mFullScreen;
                if (mFullScreen) {
                    mToolbarTop.setVisibility(View.INVISIBLE);
                } else {

                    mToolbarTop.setVisibility(View.VISIBLE);
                }
            });
        } else {
            mView = new PhotoDraweeView(this);
            GenericDraweeHierarchy hierarchy = ((PhotoDraweeView) mView).getHierarchy();
            hierarchy.setFailureImage(mDefaultDrawable, ScalingUtils.ScaleType.CENTER_CROP);
            hierarchy.setPlaceholderImage(mDefaultDrawable, ScalingUtils.ScaleType.CENTER_CROP);
            hierarchy.setActualImageScaleType(ScalingUtils.ScaleType.FIT_CENTER);
            ((PhotoDraweeView) mView).setOnViewTapListener((view, x, y) -> {
                mFullScreen = !mFullScreen;
                if (mFullScreen) {
                    mToolbarTop.setVisibility(View.INVISIBLE);
                } else {

                    mToolbarTop.setVisibility(View.VISIBLE);
                }
            });
        }
        mWholeView.addView(mView, 0, new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
        Looper.myQueue().addIdleHandler(new MyIdleHandler(this));

        mWholeView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        mBtnBack.setOnClickListener(v -> onBackPressed());

        mBtnSend.setOnClickListener(v -> {
            Intent broadcast = new Intent();
            broadcast.setAction(PICKER_ACTION_MEDIA_SURE);
            broadcast.putExtra(PICKER_EXTRA_PHOTO, mPicItem);
            sendBroadcast(broadcast);
            onBackPressed();
        });
    }

    private static final class MyIdleHandler implements MessageQueue.IdleHandler {
        private WeakReference<PictureEditPreviewActivity> mReference;

        MyIdleHandler(PictureEditPreviewActivity activity) {
            mReference = new WeakReference<>(activity);
        }

        @Override
        public boolean queueIdle() {
            if (null == mReference) return false;

            PictureEditPreviewActivity activity = mReference.get();
            if (AndroidLifecycleUtils.canLoadImage(activity)) {
                activity.init();
            }
            return false;
        }
    }

    private void init() {
        if (mView instanceof PickerScaleImageView) {
            PickerScaleImageView imageView = (PickerScaleImageView) mView;
            imageView.setMinimumTileDpi(160);

            imageView.setOnImageEventListener(new DisplayOptimizeListener(imageView));
            imageView.setMinimumScaleType(PickerScaleImageView.SCALE_TYPE_CENTER_INSIDE);
            imageView.setImage(ImageSource.uri(Uri.fromFile(new File(mPicItem.getUri()))));
        } else if (mView instanceof PhotoDraweeView) {
            PhotoDraweeView imageView = (PhotoDraweeView) mView;

            imageView.setPhotoUri(Uri.fromFile(new File(mPicItem.getUri())));
        }
    }

    private void initView() {
        mToolbarTop = findViewById(R.id.picker_preview_toolbar);
        mBtnBack = findViewById(R.id.picker_back);
        mBtnSend = findViewById(R.id.picker_sure);
        mWholeView = findViewById(R.id.picker_whole_layout);
//        mPhotoView = findViewById(R.id.picker_photo_preview);
//        mLongIv = findViewById(R.id.picker_long_photo);
        int enableColor = AttrsUtils.getTypeValueColor(this, R.attr.picker_send_color_enable);
        int disableColor = AttrsUtils.getTypeValueColor(this, R.attr.picker_send_color_disable);


        int[] colors = {
                disableColor,
                enableColor
        };
        int states[][] = new int[][]{
                new int[]{
                        -android.R.attr.state_enabled
                },
                new int[]{
                        android.R.attr.state_enabled
                }
        };

        ColorStateList colorStateList = new ColorStateList(states, colors);

        mBtnSend.setTextColor(colorStateList);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

}

File: picker/src/main/java/com/hy/picker/PictureEditPreviewActivity.kt
package com.hy.picker

import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.StateListDrawable
import android.net.Uri
import android.os.Bundle
import android.os.Looper
import android.transition.AutoTransition
import android.transition.TransitionManager
import android.view.View
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.content.ContextCompat
import com.facebook.drawee.drawable.ScalingUtils
import com.hy.picker.model.Photo
import com.hy.picker.utils.DisplayOptimizeListener
import com.hy.picker.view.ImageSource
import com.hy.picker.view.PickerScaleImageView
import kotlinx.android.synthetic.main.picker_activity_edit_preview.*
import me.relex.photodraweeview.PhotoDraweeView
import java.io.File

/**
 * Created time : 2018/8/2 8:23.
 *
 * @author HY
 */
class PictureEditPreviewActivity : PickerBaseActivity() {

    private var fullScreen = false

    private val constraintSet1 = ConstraintSet()
    private val constraintSet2 = ConstraintSet()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.picker_activity_edit_preview)

        val defaultDrawable = ContextCompat.getDrawable(this, PhotoPicker.defaultDrawable)!!

        initView()

        val intent = intent
        val picItem: Photo? = intent.getParcelableExtra(EXTRA_ITEM)

        if (picItem == null) {
            Toast.makeText(this, R.string.picker_file_error, Toast.LENGTH_SHORT).show()
            finish()
            return
        }


        val imageView: View
        if (picItem.isLong) {
            imageView = PickerScaleImageView(this)
            imageView.setOnClickListener {
                fullScreen = !fullScreen

                val autoTransition = AutoTransition()
                autoTransition.duration = 200
                TransitionManager.beginDelayedTransition(picker_whole_layout, autoTransition)
                if (fullScreen) {
                    constraintSet2.applyTo(picker_whole_layout)
                } else {
                    constraintSet1.applyTo(picker_whole_layout)
                }
            }
        } else {
            imageView = PhotoDraweeView(this)
            val hierarchy = imageView.hierarchy
            hierarchy.setFailureImage(defaultDrawable, ScalingUtils.ScaleType.CENTER_CROP)
            hierarchy.setPlaceholderImage(defaultDrawable, ScalingUtils.ScaleType.CENTER_CROP)
            hierarchy.actualImageScaleType = ScalingUtils.ScaleType.FIT_CENTER
            imageView.setOnViewTapListener { _, _, _ ->
                fullScreen = !fullScreen
                val autoTransition = AutoTransition()
                autoTransition.duration = 200
                TransitionManager.beginDelayedTransition(picker_whole_layout, autoTransition)
                if (fullScreen) {
                    constraintSet2.applyTo(picker_whole_layout)
                } else {
                    constraintSet1.applyTo(picker_whole_layout)
                }
            }
        }
        imageView.id = R.id.pickerPhotoImage


        picker_whole_layout.addView(imageView)

        val constraintSet = ConstraintSet()
        constraintSet.clone(picker_whole_layout)
        constraintSet.constrainHeight(R.id.pickerPhotoImage, 0)
        constraintSet.constrainWidth(R.id.pickerPhotoImage, 0)
        constraintSet.connect(R.id.pickerPhotoImage, ConstraintSet.TOP, R.id.pickerBackIv, ConstraintSet.BOTTOM)
        constraintSet.connect(R.id.pickerPhotoImage, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM)
        constraintSet.connect(R.id.pickerPhotoImage, ConstraintSet.START, ConstraintSet.PARENT_ID, ConstraintSet.START)
        constraintSet.connect(R.id.pickerPhotoImage, ConstraintSet.END, ConstraintSet.PARENT_ID, ConstraintSet.END)
        constraintSet.applyTo(picker_whole_layout)

        constraintSet1.clone(constraintSet)

        constraintSet2.clone(constraintSet)
        constraintSet2.clear(R.id.pickerTitleBg)
        constraintSet2.connect(R.id.pickerTitleBg, ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.TOP)
        constraintSet2.setMargin(R.id.pickerTitleBg, ConstraintSet.BOTTOM, 2)

        Looper.myQueue().addIdleHandler {
            if (canLoadImage()) {
                if (imageView is PickerScaleImageView) {
                    imageView.setMinimumTileDpi(160)

                    imageView.setOnImageEventListener(DisplayOptimizeListener(imageView))
                    imageView.setMinimumScaleType(PickerScaleImageView.SCALE_TYPE_CENTER_INSIDE)
                    imageView.setImage(ImageSource.uri(Uri.fromFile(File(picItem.uri))))
                } else if (imageView is PhotoDraweeView) {

                    imageView.setPhotoUri(Uri.fromFile(File(picItem.uri)))
                }
            }
            return@addIdleHandler false
        }


        pickerBackIv.setOnClickListener { onBackPressed() }

        pickerSure.setOnClickListener {
            val broadcast = Intent()
            broadcast.action = PICKER_ACTION_MEDIA_SURE
            broadcast.putExtra(PICKER_EXTRA_PHOTO, picItem)
            sendBroadcast(broadcast)
            onBackPressed()
        }
    }


    private fun initView() {
        val layoutParams = pickerTitleBg.layoutParams

        layoutParams.height = getStatusBarHeight() + dp(48f)
        pickerTitleBg.layoutParams = layoutParams

        val theme = PhotoPicker.theme

        pickerTitleBg.setBackgroundColor(theme.titleBgColor)
        pickerBackIv.setColorFilter(theme.backIvColor)
        pickerIndexTotalTv.setTextColor(theme.titleTvColor)

        val drawable = GradientDrawable()
        drawable.setColor(theme.sendBgColor)
        drawable.cornerRadius = dp(5f).toFloat()
//        pickerSure.background = drawable
        val sendStates = arrayOf(intArrayOf(-android.R.attr.state_enabled), intArrayOf(android.R.attr.state_enabled))

        val colorDrawable = GradientDrawable()
        colorDrawable.setColor(0x4f000000)
        colorDrawable.cornerRadius= dp(5f).toFloat()

        val layerDrawable = LayerDrawable(arrayOf(drawable, colorDrawable))

        val stateListDrawable = StateListDrawable()
        stateListDrawable.addState(intArrayOf(-android.R.attr.state_enabled),layerDrawable)
        stateListDrawable.addState(intArrayOf(android.R.attr.state_enabled),drawable)
        pickerSure.background = stateListDrawable

        val sendColors = intArrayOf(theme.sendTvColorDisable, theme.sendTvColorEnable)

        val sendColorStateList = ColorStateList(sendStates, sendColors)
        pickerSure.setTextColor(sendColorStateList)

    }

}
